### PR TITLE
Report monomorphization time errors in dead code, too

### LINKED ISSUE #107503: Const-eval errors in dead functions are optimization-dependent
I tried this code: [[playground]](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=1ffcf365aa942927bfb5f465c4dc4e99)

```rust
#![feature(inline_const)]

pub fn cause_late_error<T>() {
    const {
        panic!()
    };
}

fn main() {
    let a = 1;
    let b = 2;
    if a + b == 5 {
        cause_late_error::<()>();
    }
}
```

I expected to see this happen:

```text
error[E0080]: evaluation of `cause_late_error::<()>::{constant#0}` failed
 --> src/main.rs:7:9
  |
7 |         panic!()
  |         ^^^^^^^^ the evaluated program panicked at 'explicit panic', src/main.rs:7:9
  |
  = note: this error originates in the macro `$crate::panic::panic_2021` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn cause_late_error::<()>`
  --> src/main.rs:13:9
   |
13 |         cause_late_error::<()>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^
```

Instead, this happened: it compiled without any warnings.

### Meta

Playground nightly, 1.69.0-nightly (2023-01-30 001a77fac33f6560ff36)

### Related

- https://github.com/rust-lang/rust/issues/85836
- https://github.com/rust-lang/rust/pull/104087
- https://github.com/rust-lang/rust/issues/106617

Potential fix:

- #122568

### PR DESCRIPTION
**Pull request by oli-obk**
Before this PR

```rust

struct Foo<T>(T);

impl<T> Foo<T> {
    const BAR: () = if std::mem::size_of::<T>() == 0 {
        panic!()
        //~^ ERROR evaluation of `Foo::<()>::BAR` failed
    };
}

#[inline(never)]
fn bop<T>() {
    Foo::<T>::BAR;
}

fn fop<T>() {
    if false {
        bop::<T>();
    }
}

fn main() {
    fop::<()>();
}
```

would only error in debug mode, but pass in release mode. This behaviour [is a blocker](https://github.com/rust-lang/rust/pull/104087#issuecomment-1550208012) for the stabilization of `inline_const`.

TODO:

* [x] test performance impact
* [ ] self-review PR

Fixes https://github.com/rust-lang/rust/issues/107503

### CODE REVIEW

**Comment by oli-obk** [comment]
@rust-timer build ec8de1ebe0d698b109beeaaac83e60f4ef8bb7d1

**Comment by RalfJung** [comment]
I'm confused by this commit

>  Look only at unevaluated consts during const eval and miri.
>
> While const eval could also be optimization dependent, in practice it is not, because no MIR optimizations are run on it. Miri would still have this issue, but since it also runs the regular monomorphization, there are no issues.

Where does Miri run regular monomorphization?

Also, shouldn't evaluating already evaluated consts be cheap?

Also, what even are we potentially missing by skipping already evaluated consts? The point is to eval these consts to ensure we raise any errors that triggers. Only unevaluated consts can possibly raise any errors, so all the others already were NOPs anyway?

**Comment by oli-obk** [comment]
Evaluated consts can be function zsts or similar things that contain generics and unevaluated consts in their params. Monomorphizing and normalizing them will end up with situations where code miri never executes, but could execute and fail, will cause that failure because of unimplemented extern statics

**Comment by oli-obk** [comment]
> Where does Miri run regular monomorphization?

I thought it did via the `analysis` query, but maybe we managed to avoid all that. I'll recheck whether the newly added test passes under miri

**Comment by RalfJung** [comment]
Oh of course, the entire point is to also trigger the errors of functions that are called. But how can it be okay to skip that in Miri/CTFE?

> Monomorphizing and normalizing them will end up with situations where code miri never executes, but could execute and fail, will cause that failure because of unimplemented extern statics

I am confused. How do extern statics come in?

**Comment by oli-obk** [comment]
> How do extern statics come in?

Idk, that's what I get errors about when I don't have the last commit and I just wanted to fix it and not care about avoiding this issue in const eval or miri.

Looking at the regression, I won't put any effort into it unless that regression is considered acceptable.

For const eval the fix should be easy, but for miri it will either be very ugly or we'll need to implement every extern static in libstd, not just the ones we can actually access.

**Comment by oli-obk** [comment]
> While we want to monomorphize `required_items` and transitively evaluate all required constants, surely we don't want to actually code generate all the dead code that was optimized out?

true... I forgot about that part. That could simplify a few things.

**Comment by oli-obk** [comment]
@rust-timer build 622555246b892b174b3808f47fa66d090c8abb26

**Comment by oli-obk** [comment]
@rust-timer build 622555246b892b174b3808f47fa66d090c8abb26

**Comment by oli-obk** [comment]
OK, this took a bit of the edge off the regression, but it's still pretty bad. Expected, as we're doing extra work, but bad.

**Comment by RalfJung** [comment]
So is it time to take these numbers to the lang team and ask them if they would like these errors to be reported, now that we have an idea of the cost? Are there some things that could still be done to improve perf here?

**Comment by oli-obk** [comment]
Nominating for lang team discussion. See the changed test (or the example in the main post of this PR) for an example of code that compiles in release mode, but not in debug mode. This PR changes that to also error in release mode, at the cost of performing more work in both release and debug mode, as hypothetically debug mode can also make dead code disappear.

Performance in incremental tests regresses up to 25% in extreme cases. Full builds regress up to 4% (except for the `projection-caching` test, which also regresses by 25%, but that's an artificial stress test).

Note that all of the regressions are for cases where we *don't* have an issue, but we also don't know this ahead of time. It's unlikely we can do much better here, as we are already avoiding duplicate work and ensuring that we still don't generate LLVM IR for dead code, even if LLVM could raise an error there in debug mode, but not release mode (a [previous performance run](https://github.com/rust-lang/rust/pull/112879#issuecomment-1600664132) that did that was much worse).

**Comment by tmandry** [comment]
To make sure we are talking about the same thing, let me state my understanding: This causes all type-level instantiations to have MIR built and const expressions evaluated at compile time. Previously instantiations inside expressions that were optimized out could be skipped, and I would expect this to happen primarily in release mode.

But I'm confused why the instruction counts for debug mode look _almost exactly like_ the counts for release mode, as I would expect the difference to be much larger for release mode (where more code is optimized out) and almost nonexistent for debug mode.

**Comment by tmandry** [comment]
Discussed in today's [lang team meeting](https://hackmd.io/tjdvwMJzSyWs0sjcl89p6Q#%E2%80%9CReport-monomorphization-time-errors-in-dead-code-too%E2%80%9D-rust112879).

Based on our understanding of the PR (which roughly tracks the understanding I posted in the above comment), we agreed that this was probably a good change in spite of the significant performance regression.

## Not establishing a policy

While we think this change is worthwhile, we don't want landing this PR to be interpreted as a policy that we will evaluate all type-level instantiated code going forward. We should consider that as a separate proposal to be FCP'd.

## Crater run

We think there should be a crater run to detect unexpected or widesread regressions, though if there are differences in modes that may be tricky. Do you have any thoughts or suggestions on this, @oli-obk?

Probably the most sensitive thing would be to run crater in release mode, which is probably already quite slow, but it would _also_ be useful to know if any regressed crates also fail in debug mode. The most interesting cases would be crates that regress in one mode that do not also fail in the other mode, if any.

## Performance

On the performance side, perhaps @rust-lang/wg-compiler-performance will have ideas on how we could minimize the impact?

As I write this, another idea that comes to mind would be to land this change under a `-Z` flag so people could test on their code bases before turning it on by default.

**Comment by nnethercote** [comment]
> On the performance side, perhaps @rust-lang/wg-compiler-performance will have ideas on how we could minimize the impact?

If you do a lot more work it's going to take longer :(

**Comment by tmandry** [comment]
@rfcbot fcp merge

We agreed in last week's lang team meeting to FCP this. My understanding is that this removes all(?) errors that are dependent on optimization flags (e.g. debug vs release), but does not tackle the difference between `check` and `build`.

My [above comment](https://github.com/rust-lang/rust/pull/112879#issuecomment-1641214618) still applies to what's being FCP'd: This FCP does not itself establish a policy that we will commit to this kind of behavior going forward. A crater run would be informative in terms of both build failure regressions and performance (if we are set up to collect performance data).

---

I'm still trying to understand why there were regressions on debug builds too. I'm guessing it's because we optimized dead code out even in debug mode? That would suggest that there are going to be ecosystem crates which used to build but do not build anymore, in either debug or release.

**Comment by joshtriplett** [comment]
> > On the performance side, perhaps @rust-lang/wg-compiler-performance will have ideas on how we could minimize the impact?
> 
> If you do a lot more work it's going to take longer :(

@nnethercote That's true, but would it be possible to optimize this case and get some of the performance back, perhaps by recognizing the "dead code" case and reducing the amount of work done that *isn't* require to report the error?

**Comment by scottmcm** [comment]
🤞 that we can add some new language features to *intentionally* control some of this, and thus still be able to pre-optimize-out some of this in a way that's less subject to the whims of random mir-opt changes.

File: compiler/rustc_codegen_cranelift/src/constant.rs [hunk 1]
```diff
@@ -35,9 +35,11 @@ impl ConstantCx {
 
 pub(crate) fn check_constants(fx: &mut FunctionCx<'_, '_, '_>) -> bool {
     let mut all_constants_ok = true;
-    for constant in &fx.mir.required_consts {
-        if eval_mir_constant(fx, constant).is_none() {
-            all_constants_ok = false;
+    for (item, _) in &fx.mir.required_items {
+        if let rustc_middle::mir::MonoItem::Const(constant) = item {
```
Comment (RalfJung) [inline compiler/rustc_codegen_cranelift/src/constant.rs:39]: We have this logic a couple of times. Might be worth it to have a `required_consts` method on `mir::Body`?

File: compiler/rustc_const_eval/src/interpret/eval_context.rs [hunk 2]
```diff
@@ -703,10 +703,19 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+        for (ct, span) in &body.required_items {
+            if let mir::MonoItem::Const(ct) = ct {
+                match ct.literal {
+                    mir::ConstantKind::Ty(c) => match c.kind() {
+                        ty::ConstKind::Unevaluated(_) => {}
+                        _ => continue,
+                    },
+                    mir::ConstantKind::Unevaluated(_, _) => {}
+                    mir::ConstantKind::Val(_, _) => continue,
+                }
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/eval_context.rs:715]: What's this about? The codegen backends don't do this. At least this needs a comment explaining why we are skipping some consts.

**Comment by nnethercote** [comment]
> @nnethercote That's true, but would it be possible to optimize this case and get some of the performance back, perhaps by recognizing the "dead code" case and reducing the amount of work done that _isn't_ require to report the error?

Well, maybe. But ideas are cheap and execution is hard. Look at polymorphism: it's a good idea and conceptually simple and could give big wins and is really, really hard to get working in practice. So I strongly recommend that this decision should *not* be driven by any assumption that the performance will be won back cheaply.

As a @rust-lang/wg-compiler-performance member it's really dispiriting to see something like a year's worth of performance improvements possibly undone in a single PR. All the low-hanging performance fruit is gone, so gaining back that performance will likely be more than a year's additional work.

**Comment by joshtriplett** [comment]
@nnethercote :+1: for making sure the signoff here is not contingent on finding ways to mitigate the performance. (I'm re-nominating this for lang so that we can confirm that. @scottmcm, @tmandry, are either of your checkboxes contingent on assumptions that we will get this specific performance loss back?)

A 4-6% performance regression is awful. The only reason we're currently willing to reluctantly live with it is that *without* this change, compiler optimizations are potentially part of the compiler's stable surface area, and modifying an optimization could cause existing code to fail to compile. *That* is the outcome that's so unpalatable as to be worth considering this.

If there's any *other* way we can avoid having compiler optimizations become part of Rust's stability guarantees, even a way that has *other* substantial tradeoffs, I think we'd be extremely interested in considering it.

**Comment by RalfJung** [comment]
All the incr-unchanged regressions are also strange, if nothing changed caching should be as effective as before, no? Can someone generate the cachegrind diff, e.g. for [this benchmark](https://perf.rust-lang.org/detailed-query.html?commit=622555246b892b174b3808f47fa66d090c8abb26&base_commit=3c554f5cb4965a08f2a2b45ab4bba321a5f249fa&benchmark=cargo-0.60.0-debug&scenario=incr-unchanged)? E.g. there's a 25% incremental loading delta for optimized_mir, which is odd since we shouldn't run any more optimizations than before definitely not in a debug build...

**Comment by joshtriplett** [comment]
@rfcbot concern incremental

Filing a concern until the incremental cases @RalfJung is identifying get root-caused.

**Comment by oli-obk** [comment]
I have not found the time to look at where the regression is coming from. I will not be able to get to this before January, so if someone could pick this up this PR to speed up getting this (and then `const` blocks) landed, that would be awesome.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

I'm going to unnominate this, as there's nothing more for T-lang to discuss until this is investigated.  For anyone who picks this up, please renominate this after posting the investigation (or fix).

**Comment by the8472** [comment]
Based on conversation in https://github.com/rust-lang/rfcs/issues/3582 I have modified the PR example by moving thea dead branch in const eval instead of a regular function.

```rust
struct Foo<T, const N: usize>(T);

impl<T, const N: usize> Foo<T, N> {
    const BAR: () = if N == 0 {
        panic!()
    };
}

struct Invoke<T, const N: usize>(T);

impl<T, const N: usize> Invoke<T, N> {
    const FUN: fn() = if N != 0 {
        || Foo::<T, N>::BAR
    } else {
        || {}
    };
}

fn main() {
    Invoke::<(), 0>::FUN();
}
```

This passes on stable with and without optimizations. Would this PR cause it to fail?

**Comment by RalfJung** [comment]
I started experimenting a bit in https://github.com/rust-lang/rust/pull/122258 to see which step costs how much perf.

But I think I also better understand now how const-eval errors behave during monomorphization. There are actually two stages of monomorphization: collect all the things we want to codegen, and then codegen them all. These seem to happen sequentially, and if there is an error during collection we never go to codegen. Const-eval errors are triggered during collection (the collector triggers the const-eval queries).

Our goal (to unblock `const {}`) is just to detect const-eval errors in dead code, not errors that happen later during codegen. I think this PR will make dead functions fully monomorphized, i.e., they are actually passed to codegen -- when all that we really want is to walk them and trigger evaluation of all consts that we can find (recursively). Perf will probably look a lot better if we can avoid sending these functions to codegen. The monomorphization collection just has a single walking mode though, "walk everything, evaluate consts and add functions to the codegen list".

I think we either want to change monomorphization to walk all `required_items` without sending them to codegen (only send them to codegen if we also find the same item in the MIR), or we want a separate traversal that does the recursive search for all consts. That separate traversal would be very similar to the monomorphization traversal though so this doesn't sound great, we'd probably duplicate a lot of effort and code. Making monomorphization keep track of whether we are currently visiting "for const eval only" or "for codegen" could also become a bit messy but still seems more promising IMO.

I haven't looked into monomorphization before so take this with a bunch of salt.

**Comment by RalfJung** [comment]
@the8472 We'd have to decide whether dead code in const is monomorphized. Currently it is not, and I think this PR doesn't change that. This is not the result of optimizations; the collector simply does not walk the MIR of a const and const-eval only monomorphizes functions that actually get called. We could make const-eval iterate over `required_items` and recursively search them for consts to evaluate, then your example would fail to build. I think whether or not we want that is largely orthogonal to this PR, and it shouldn't be on the critical path for `const {}` as it is already not optimization-dependent. So please open a new issue. :)

**Comment by RalfJung** [comment]
I think this is superseded by https://github.com/rust-lang/rust/pull/122568. I managed to get the perf overhead down to 3% for the worst case, and it almost exclusively affects incr builds (which, in the benchmark suite, are "incr with no or almost no change").

**Pull request closed without merging**